### PR TITLE
Generate Docker Image for Reports

### PR DESCRIPTION
**Pull request by iainjwhiteside**
## Description of change

Introduces some Bazel functionality to build and push our generic reports functionality as a docker image to be directly used as a metrics build.

## Guide to reproduce test results.

N/A

## Checklist:
- [ ] I have self-reviewed this change.
- [ ] I have tested this change.
- [ ] This change is covered by tests that are already landed, or in this PR.

### CODE REVIEW

File: .github/workflows/build-push-default-reports-build.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  push:
+    branches: 
+    - main
+
+permissions:
+  id-token: write
+  contents: read
+
+env:
+  REPO_NAME: customer-test-images
```
Comment (iainjwhiteside) [inline .github/workflows/build-push-default-reports-build.yml:12]:
```suggestion
  REPO_NAME: open-metrics-builds
```

File: .github/workflows/build-push-default-reports-build.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    branches: 
+    - main
+
+permissions:
+  id-token: write
+  contents: read
+
+env:
+  REPO_NAME: open-metrics-builds
+  REGISTRY_NAME: 909785973729.dkr.ecr.us-east-1.amazonaws.com
```
Comment (iainjwhiteside) [inline .github/workflows/build-push-default-reports-build.yml:13]:
```suggestion
  REGISTRY_NAME: public.ecr.aws
```

File: resim/metrics/default-report-metrics/entrypoint.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+async def main() -> None:
+    report_metrics_config = pathlib.Path(REPORT_METRICS_CONFIG_PATH)
+    is_report_metrics_mode = report_metrics_config.is_file()
+    if is_report_metrics_mode:
+        # Call the default_report_metrics main function:
+        await default_report_metrics_main()
+    else:
+        # Otherwise, we can't do anything: exit 1
+        print("Metrics Build only compatible with reports")
+        sys.exit(1)
```
Comment (mikebauer) [inline resim/metrics/default-report-metrics/entrypoint.py:21]: [nit] I lean towards just doing this in the default report metrics file, but I won't block on it. Eventually, I hope that file can look much more like this one once we've abstracted out and tested most of it.

File: resim/metrics/default-report-metrics/entrypoint.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/python3
```
Comment (mikebauer) [inline resim/metrics/default-report-metrics/entrypoint.py:1]: copyright notice missing

File: resim/metrics/default-report-metrics/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+RUN apt -y install wget python3 python3-pip git
+
+COPY ../../../requirements.txt /
+RUN pip3 install -r requirements.txt
+
+COPY src/ /workspaces/
+
+WORKDIR /usr/bin
+
+ENTRYPOINT /workspaces/entrypoint.py
```
Comment (mikebauer) [inline resim/metrics/default-report-metrics/Dockerfile:13]: We should be able to remove this file.

File: resim/metrics/default-report-metrics/BUILD [hunk 6]
```diff
@@ -0,0 +1,39 @@
+load("@aspect_bazel_lib//lib:tar.bzl", "tar")
```
Comment (mikebauer) [inline resim/metrics/default-report-metrics/BUILD:1]: copyright notice

File: resim/metrics/default-report-metrics/BUILD [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+oci_tarball(
+    name = "default_report_metrics_build_tarball",
+    image = ":default_report_metrics_build",
+    repo_tags = ["default-report-metrics-build:latest"],
+)
+
+oci_push(
+    name = "default_report_metrics_build_push",
+    image = ":default_report_metrics_build",
+    repository = "public.ecr.aws/resim/open-metrics-builds",  # Test!!
```
Comment (mikebauer) [inline resim/metrics/default-report-metrics/BUILD:38]: Intentional comment?

File: WORKSPACE.bzlmod [hunk 8]
```diff
@@ -0,0 +1,10 @@
+load("@rules_oci//oci:pull.bzl", "oci_pull")
+
+oci_pull(
+    name = "ubuntu",
+    image = "ubuntu:jammy",
+    platforms = [
+        "linux/arm64/v8",
+        "linux/amd64",
+    ],
+)
```
Comment (mikebauer) [inline WORKSPACE.bzlmod:10]: We should be able to remove this file I think.

File: .github/workflows/build-push-default-reports-build.yml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+          SHA: ${{ github.sha }}
+        run: bazel build //resim/metrics/default-report-metrics:default_report_metrics_build
+
+      - name: Push
+        if: ${{ github.ref == 'refs/heads/main' }}
+        env:
+          REPO_NAME: ${{ env.REGISTRY_NAME }}/${{ env.REPO_NAME }}
+          IMAGE_VERSION: ${{ steps.meta.outputs.version }}
+          SHA: ${{ github.sha }}
+        run: bazel run //:default_report_metrics_build_push -- --repository "$REPO_NAME" --tag "$IMAGE_VERSION" --tag "sha-${SHA}"
```
Comment (mikebauer) [inline .github/workflows/build-push-default-reports-build.yml:60]: I think we need to be careful about how we do this. As it currently stands, it seems like there's nothing in the full URI of the image being pushed to identify it as a report metrics build and if we push other images their tags will collide unless we differentiate. Maybe `--tag "report_metrics_${IMAGE_VERSION}"` or something like that?

**Review by mikebauer — CHANGES_REQUESTED** [review]
Requesting a few minor changes on this one. Happy to look again first thing in the morning.

File: resim/metrics/default-report-metrics/entrypoint.py (hunk 3 above)
Comment (iainjwhiteside) [reply resim/metrics/default-report-metrics/entrypoint.py:21]: SGTM

File: resim/metrics/default_report_metrics_build/BUILD [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+oci_tarball(
+    name = "default_report_metrics_build_tarball",
+    image = ":default_report_metrics_build",
+    repo_tags = ["default-report-metrics-build:latest"],
+)
+
+oci_push(
+    name = "default_report_metrics_build_push",
+    image = ":default_report_metrics_build",
+    repository = "public.ecr.aws/resim/open-metrics-builds",
```
Comment (iainjwhiteside) [inline resim/metrics/default_report_metrics_build/BUILD:34]:
```suggestion
    repository = "public.ecr.aws/resim/open-metrics-builds/default-reports-build",
```

File: .github/workflows/build-push-default-reports-build.yml (hunk 9 above)
Comment (iainjwhiteside) [reply .github/workflows/build-push-default-reports-build.yml]: good point. moved to using namespacing to keep the tag clean

File: resim/metrics/default_report_metrics_build/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+
+RUN apt update
+RUN apt -y install wget python3 python3-pip git
+
+COPY ../../../requirements.txt /
+RUN pip3 install -r requirements.txt
+
+COPY src/ /workspaces/
+
+WORKDIR /usr/bin
```
Comment (mikebauer) [inline resim/metrics/default_report_metrics_build/Dockerfile:11]: I think this file can be removed.

**Review by mikebauer — APPROVED** [review]
Thanks for addressing my comments!

**Pull request merged**